src/rb_sys.rs

```diff
@@ -24,45 +24,119 @@
+    ///
+    /// ```
+    /// # let _cleanup = unsafe { magnus::embed::init() };
+    ///
+    /// use magnus::{RString, Value, rb_sys::FromRawValue};
+    ///
+    /// let raw_value = unsafe { rb_sys::rb_str_new("foo".as_ptr() as *mut _, 3) };
+    ///
+    /// assert_eq!(unsafe { Value::from_raw(raw_value) }.to_string(), "foo");
+    /// ```
+    unsafe fn from_raw(value: VALUE) -> Self;
 }
 
-/// Convert [`rb_sys::VALUE`] to [`magnus::Value`](Value).
-///
-/// # Safety
-///
-/// You must only supply a valid [`VALUE`] obtained from [rb-sys](rb_sys) to
-/// this function. Using a invalid [`Value`] produced from this function will
-/// void all saftey guarantees provided by Magnus.
-pub unsafe fn value_from_raw(val: VALUE) -> Value {
-    Value::new(val)
+impl AsRawValue for Value {
+    fn as_raw(self) -> VALUE {
+        self.as_rb_value()
+    }
 }
 
-/// Convert [`rb_sys::ID`] to [`magnus::value::Id`](Id).
-///
-/// # Safety
-///
-/// You must only supply a valid [`ID`] obtained from [rb-sys](rb_sys) to this
-/// function. Using a invalid [`Id`] produced from this function will void all
-/// saftey guarantees provided by Magnus.
-pub unsafe fn id_from_raw(id: ID) -> Id {
-    Id::new(id)
+impl FromRawValue for Value {
+    unsafe fn from_raw(val: VALUE) -> Value {
+        Value::new(val.into())
+    }
+}
+
+/// Trait to convert a [`Id`] to a raw [`ID`].
+pub trait AsRawId {
+    /// Convert [`magnus::value::Id`](Id) to [`rb_sys::ID`](ID).
+    ///
+    /// ```
+    /// # let _cleanup = unsafe { magnus::embed::init() };
+    ///
+    /// use magnus::{Symbol, value::Id, rb_sys::{AsRawId, FromRawId}};
+    ///
+    /// let foo: Id = Symbol::new("foo").into();
+    /// let raw = foo.as_raw();
+    /// let from_raw_val: Symbol = unsafe { Id::from_raw(raw) }.into();
+    ///
+    /// assert_eq!(from_raw_val.inspect(), ":foo");
+    /// ```
+    fn as_raw(self) -> ID;
+}
+
+/// Trait to convert from a raw [`ID`] to an [`Id`].
+pub trait FromRawId {
+    /// Convert [`rb_sys::ID`](ID) to [`magnus::value::Id`](ID).
+    ///
+    /// # Safety
+    ///
+    /// You must only supply a valid, non-zero [`ID`] obtained from [rb-sys](rb_sys) to this
+    /// function. Using a invalid [`Id`] produced from this function will void all
+    /// saftey guarantees provided by Magnus.
+    ///
+    /// ```
+    /// # let _cleanup = unsafe { magnus::embed::init() };
+    ///
+    /// use magnus::{Symbol, value::Id, rb_sys::{FromRawId, AsRawId}};
+    /// use std::convert::TryInto;
+    ///
+    /// let foo: Id = Symbol::new("foo").into();
+    /// let from_raw_val: Symbol = unsafe { Id::from_raw(foo.as_raw()) }.into();
+    ///
+    /// assert_eq!(from_raw_val.inspect(), ":foo");
+    /// ```
+    unsafe fn from_raw(id: ID) -> Self;
+}
+
+impl AsRawId for Id {
+    fn as_raw(self) -> ID {
+        self.as_rb_id()
+    }
+}
+
+impl FromRawId for Id {
+    unsafe fn from_raw(id: ID) -> Id {
+        debug_assert!(!Value::is_immediate(transmute::<ID, Value>(id)));
```

Comment:
I'm not sure this check is legit, I put together a quick test, which fails:

`tests/from_raw_id.rs`
``` rust
use magnus::{
    class,
    rb_sys::{AsRawId, FromRawId},
    value::Id,
    Module, RArray, RClass, Symbol,
};

#[test]
fn roundtrip_all_symbols() {
    let _cleanup = unsafe { magnus::embed::init() };

    let sym_class: RClass = class::object().const_get("Symbol").unwrap();
    let symbols: RArray = sym_class.funcall("all_symbols", ()).unwrap();

    for sym in symbols.each() {
        let sym: Symbol = sym.unwrap().try_convert().unwrap();
        let id: Id = sym.into();

        println!("{:?} - {}", sym, id.as_raw());

        unsafe {
            Id::from_raw(id.as_raw());
        }
    }
}
```

outputs
``` sh
~/magnus> cargo test --test from_raw_id -- --nocapture
   Compiling magnus v0.3.1 (~/magnus)
    Finished test [unoptimized + debuginfo] target(s) in 0.21s
     Running tests/from_raw_id.rs (~/magnus/target/debug/deps/from_raw_id-cb31cc4bb7278fdc)

running 1 test
:! - 33
thread 'roundtrip_all_symbols' panicked at 'assertion failed: !Value::is_immediate(transmute::<ID, Value>(id))', src/rb_sys.rs:136:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test roundtrip_all_symbols ... FAILED

failures:

failures:
    roundtrip_all_symbols

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.04s

error: test failed, to rerun pass '--test from_raw_id'
```